Single typing file

guybedford opened the pull request:
This fixes up the main Rollup API typings to be a single curated `types.d.ts` file, which other internal typings then import from.

It does mean maintaining some exported dependency typing boundaries a little. The only real issue is typings of AST Nodes not passing through, but this doesn't seem to much of an issue (I've just aliased these any for now).

What's also quite nice is the typing catches the difference between code splitting and non code splitting cases an outputs, so gives correct IDE hints through the build.

Fixes #2008 and replaces #2094.

lukastaegert commented on package.json:
```diff
@@ -5,7 +5,7 @@
   "main": "dist/rollup.js",
```
This change also means that we can–and should–move all `dependencies` back into `devDependencies` so that rollup is now back to its small and lean footprint!

lukastaegert commented:
> Fixes #2008 and replaces #2094 

I believe it is the other way around?

lukastaegert commented:
@diervo, @KingHenne, @jthoms1, @alan-agius4, @chadhietala as you have taken some interest in rollup's typings before, maybe you could take a look by installing rollup via

```
npm install rollup/rollup#typing-file
```

and see if this works for you? If there are no hard objections I would love to see this released soon.

alan-agius4 commented on src/rollup/types.d.ts:
```diff
@@ -0,0 +1,343 @@
+	pluginCode?: string;
+}
+
+export type VERSION = string;
```
This doesn't look right, AFAIK version return a string value and thus it should be a `const` and not `type`

The usage I believe it should be;

```ts
import { VERSION } from 'rollup';
const rollupVersion = VERSION:
```

guybedford replied:
Thanks

alan-agius4 commented on src/rollup/index.ts:
```diff
@@ -1,204 +1,24 @@
 import { createAddons } from '../utils/addons';
 import commondir from '../utils/commondir';
-
-export const VERSION = '<@VERSION@>';
```
has this been removed on purpose?

guybedford replied:
Well caught - yes I removed this one by mistake.

alan-agius4 commented on src/rollup/types.d.ts:
```diff
@@ -0,0 +1,343 @@
+	perf?: boolean;
+
+	// deprecated
+	entry?: string;
```
just a small note, for `deprecations` it's suggest to use /** @deprecated */ on top of each property that is deprecated.

Reason being is that if you have `tslint` it can warn you that you are using a deprecated property, so do the `IDE's` or any code analyse tool such as SonarCube.

alan-agius4 commented:
Good stuff @guybedford! 

I tried them out and they do work, Just felt a couple of comment. 

Just a small side node `@types/node` will still need to be installed as a devDependency by the consumers of rollup typings.

alan-agius4 commented:
Good stuff @guybedford! 

I tried them out and they do work, Just felt a couple of comment. 

Just a small side note `@types/node` will still need to be installed as a `devDependency` by the consumers of rollup typings due to the dependency on `events` module. As otherwise it will cause `node_modules/rollup/dist/rollup.d.ts(1,30): error TS2307: Cannot find module 'events'.`

alan-agius4 commented:
Good stuff @guybedford! 

I tried them out and they do work, Just felt a couple of comment. 

Just a small side note `@types/node` will still need to be installed as a `devDependency` by the consumers of rollup typings due to the dependency on `events` module. As otherwise it will cause `node_modules/rollup/dist/rollup.d.ts(1,30): error TS2307: Cannot find module 'events'.`

What I don't like in all fairness the fact that the types for `source-map` are copied in the `rollup` typings, which means. That now, you need to maintain both your typings and `source-maps` in this repo, So that means that you need to keep the typings of `rollup` in sync with your code base. (Since know they are not automatically generated via `TS`) and with `source-maps` interfaces if they do any change.

KingHenne commented on package.json:
```diff
@@ -22,7 +22,7 @@
     "ci": "npm run test-coverage && codecov < coverage/coverage-remapped.lcov",
     "prebuild": "rm -rf dist",
-    "build": "git rev-parse HEAD > .commithash && rollup -c && tsc -p tsconfig-types.json && chmod a+x bin/rollup",
+    "build": "git rev-parse HEAD > .commithash && rollup -c && cp src/rollup/types.d.ts dist/rollup.d.ts && chmod a+x bin/rollup",
```
I think `tsconfig-types.json` is now obsolete and can be removed.

KingHenne commented on package.json:
```diff
@@ -5,7 +5,7 @@
   "module": "dist/rollup.es.js",
   "jsnext:main": "dist/rollup.es.js",
-  "typings": "dist/typings/node-entry.d.ts",
+  "typings": "dist/rollup.d.ts",
```
Since there are no files in `dist/typings` and the files in `typings` are not relevant anymore for the published artifact:
1. The script `"pretest:typescript"` can be simplified to `"rm -rf test/typescript/dist && cp -r dist test/typescript/"` and
2. `"typings/package.json.d.ts"` can be removed from `"files"`.

KingHenne commented on src/rollup/types.d.ts:
```diff
@@ -0,0 +1,343 @@
+import { EventEmitter } from 'events';
+
+// ast node types just any, unless acorn public dependency added
```
Hm, that's kind of a step back. As a plugin author I'm loosing information here since `Program` and `Node` are referenced (indirectly) in some of the hooks.

guybedford commented:
I've removed the unused package.json dependencies and scripts, and updated the dependencies to just reference types/estree and types/node generically.

Thanks all for feedback so far - please try it out again and let me know any further comments.

KingHenne commented on src/rollup/types.d.ts:
```diff
@@ -1,8 +1,7 @@
 
-// ast node types just any, unless acorn public dependency added
-export type Program = any;
-export type Node = any;
```
👍

guybedford commented:
Thanks @KingHenne, I've fixed those remaining type errors.

KingHenne commented on src/Module.ts:
```diff
@@ -148,7 +134,7 @@ export default class Module {
 
 	ast: Program;
-	private astClone: Program;
+	private astClone: ESTree.Program;
```
This confuses me. Shouldn't `ast` and `astClone` have the same type?

guybedford replied:
There are really three things here not two - `ast`, `astClone` and `enhancedAst`. `ast` doubles as both EStree and the enhanced form, while the cache API for ast and astClone relies only on the EStree interface.

KingHenne replied:
Interesting. Should `ast` be defined as `Program & ESTree.Program` then?

guybedford replied:
Program is a superset of ESTree.Program, so that would be the same thing here I think.

KingHenne replied:
I think that is the confusing part to me, because the `ast` that is passed to `setSource` is defined as `ESTree.Program` (via `ModuleJSON`). So assigning it to a field of type `Program`, which is supposed to be a superset, is technically an error then which is hidden by the [type cast](https://github.com/rollup/rollup/pull/2108/files/794bd7464a7bb90c84d4d18b47f0296247561105#diff-356410863c3a18aee7d09896d52d6b08R199), right?

guybedford replied:
The way to think about this is that `Module.ast` when first assigned is an `ESTree.Program` but transitions to a `Rollup.Program` during the enhance call (initiated by analyse). When `Module.ast` is exported, cloned or copied, even though it is a `Rollup.Program` it is only needed to be a `ESTree.Program` through the cache interfaces, as later on enhance will be called again on it.

That's why I say it's really two things - `Module.enhancedAst` and `Module.ast`.

lukastaegert replied:
Note this problem is only temporary as I am currently working on a branch that will finally separate these two types. In the end, `Module.ast` will be only a `Program` while the `Program` constructor will receive an `ESTree.Program` as parameter.

KingHenne replied:
Hm, that's tricky. In my opinion, it should really be `ESTree.Program | Program` then and consequently all five compiler errors that result from this change should add a guard so that this won't lead to a runtime error in the future when someone changes how and when the different methods in `Module` are called for some reason, e.g.:

```ts
if ('bind' in node) {
	node.bind();
}
```

or:

```ts
if (!('bind' in node)) {
	throw new Error('Unexpected AST node found. ...');
}
```

KingHenne commented on src/Module.ts:
```diff
@@ -217,12 +196,12 @@ export default class Module {
 			// TODO what happens to comments if AST is provided?
-			this.ast = <any>tryParse(this, this.graph.acornParse, this.graph.acornOptions);
-			this.astClone = clone(this.ast);
+			this.ast = <Program>tryParse(this, this.graph.acornParse, this.graph.acornOptions);
```
Continuing my previous question: Isn't `this.graph.acornParse` returning an `ESTree.Program`?

guybedford replied:
Yes, but then we mutate this later through `enhance`, and TypeScript doesn't support mutating a type!

guybedford commented on package.json:
```diff
@@ -52,19 +52,13 @@
-    "require-relative": "^0.8.7",
-    "rollup-pluginutils": "^2.0.1"
+    "@types/estree": "0.0.38",
+    "@types/node": "*"
```
Thanks, I missed the devDependency here actually (and for some reason my npm wasn't giving a warning on this).

Well EventEmitter has been there since the very start in Node so yes, any version. The star is so that it can work alongside any version a user may install.

lukastaegert commented:
Sorry for the wait, will merge this today.
